src/storage.rs

```diff
@@ -96,167 +151,230 @@ pub trait Storage {
+        self.handle_conf_changes_after_append(&ents[i..]);
     }
 
-    /// Overwrites the contents of this Storage object with those of the given snapshot.
-    pub fn apply_snapshot(&mut self, snapshot: Snapshot) -> Result<()> {
-        // handle check for old snapshot being applied
-        let index = self.snapshot.get_metadata().get_index();
-        let snapshot_index = snapshot.get_metadata().get_index();
-        if index >= snapshot_index {
-            return Err(Error::Store(StorageError::SnapshotOutOfDate));
+    fn handle_conf_changes_after_append(&mut self, ents: &[Entry]) {
+        if ents.is_empty() {
+            return;
+        }
+        // Revert some configuration changes.
+        let recover_prior = ents[0].get_index();
+        for i in (0..self.raft_state.conf_states.len()).rev() {
+            if self.raft_state.conf_states[i].index < recover_prior {
+                break;
+            }
+            self.raft_state.conf_states.pop();
         }
 
-        let mut e = Entry::new();
-        e.set_term(snapshot.get_metadata().get_term());
-        e.set_index(snapshot.get_metadata().get_index());
-        self.entries = vec![e];
-        self.snapshot = snapshot;
-        Ok(())
-    }
-
-    /// Makes a snapshot which can be retrieved with snapshot() and
-    /// can be used to reconstruct the state at that point.
-    /// If any configuration changes have been made since the last compaction,
-    /// the result of the last apply_conf_change must be passed in.
-    pub fn create_snapshot(
-        &mut self,
-        idx: u64,
-        cs: Option<ConfState>,
-        pending_membership_change: Option<ConfChange>,
-        data: Vec<u8>,
-    ) -> Result<&Snapshot> {
-        if idx <= self.snapshot.get_metadata().get_index() {
-            return Err(Error::Store(StorageError::SnapshotOutOfDate));
+        for e in ents
+            .iter()
+            .filter(|e| e.get_entry_type() == EntryType::EntryConfChange)
+        {
+            let mut cc = ConfChange::new();
+            cc.merge_from_bytes(e.get_data()).unwrap();
+            let last = self.raft_state.conf_states.last().unwrap();
+            let base = last.conf_state.clone();
+            let index = e.get_index();
+            let cs = ConfStateWithIndex::apply_conf_change(base, index, cc);
+            self.raft_state.conf_states.push(cs);
+        }
+    }
+
+    fn inner_first_index(&self) -> u64 {
+        let first_index = self.entries[0].get_index();
+        if self.snapshot_metadata.get_index() == first_index {
+            return first_index + 1;
         }
+        first_index
+    }
+
+    fn inner_last_index(&self) -> u64 {
+        self.entries.last().unwrap().get_index()
+    }
 
-        let offset = self.entries[0].get_index();
-        if idx > self.inner_last_index() {
+    /// Overwrites the contents of this Storage object with those of the given snapshot.
+    pub fn apply_snapshot(&mut self, mut snapshot: Snapshot) {
```

Comment:
Oh, these are safe now?  What if a snapshot is out of date?

Reply:
Panic. I will document it.

Reply:
Is it possible to keep it a `Result`? I would prefer to avoid any non-critical panics.